src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+        private void MarkTrustedChainAsAssumedValid(ChainedHeader chainedHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            ChainedHeader current = chainedHeader;
+
+            while (this.HeaderWasNotMarkedAsValidated(current))
+            {
+                current.BlockValidationState = ValidationState.AssumedValid;
+                current = current.Previous;
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// The header is assumed to be valid, the header and all of its previous headers will be marked as <see cref="ValidationState.AssumedValid"/>.
+        /// If the header's cumulative work is better then <see cref="IChainState.ConsensusTip"/> the header and all its predecessors will be marked with <see cref="BlockDataAvailabilityState.BlockRequired"/>.
+        /// </summary>
+        /// <param name="assumedValidHeader">The header that is assumed to be valid.</param>
+        /// <param name="latestNewHeader">The last header in the list of presented new headers.</param>
+        private ConnectedHeaders HandleAssumedValidHeader(ChainedHeader assumedValidHeader, ChainedHeader latestNewHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(assumedValidHeader), assumedValidHeader, nameof(latestNewHeader), latestNewHeader);
+
+            ChainedHeader bestTip = this.chainState.ConsensusTip;
+            ConnectedHeaders connectedHeaders = new ConnectedHeaders() {Consumed = latestNewHeader};
+
+            if (latestNewHeader.ChainWork > bestTip.ChainWork)
+            {
+                this.logger.LogDebug("Chained header '{0}' is the tip of a chain with more work than our current consensus tip.", latestNewHeader);
+
+                connectedHeaders = this.MarkBetterChainAsRequired(latestNewHeader);
+            }
+
+            this.MarkTrustedChainAsAssumedValid(assumedValidHeader);
+
+            this.logger.LogTrace("(-):{0}", connectedHeaders);
+            return connectedHeaders;
+        }
+
+        /// <summary>
+        /// When a header is before the last checkpoint it will be asked to be marked as AssumedValid and we'll ask to download it.
+        /// </summary>
+        private ConnectedHeaders HandleCheckpointsHeader(ChainedHeader chainedHeader, ChainedHeader latestNewHeader, CheckpointInfo checkpoint)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}',{4}.{5}:'{6}')", nameof(chainedHeader), chainedHeader, nameof(latestNewHeader), latestNewHeader, nameof(checkpoint), nameof(checkpoint.Hash), checkpoint.Hash);
+
+            if (checkpoint.Hash != chainedHeader.HashBlock)
+            {
+                this.logger.LogDebug("Chained header '{0}' does not match checkpoint '{1}'.", chainedHeader, checkpoint.Hash);
+                this.logger.LogTrace("(-)[INVALID_HEADER_NOT_MATCHING_CHECKPOINT]");
+                throw new InvalidHeaderException();
```

Comment:
There is little bit of fuck up here. It is somehow undefined when we have been provided assume valid header on a chain that violates the checkpoint. It is quite a case which will never be a problem, so I think we are fine if we just say that the behaviour is undefined if assume valid header defines the chain that violates the checkpoint. Maybe we should put such a warning into the comment of assume valid hash definition.